pkg/promtail/targets/awss3target.go

```diff
@@ -0,0 +1,228 @@
+	flags.BoolVar(&cfg.Stdin, "stdin", false, "Set to true to pipe logs to promtail.")
+}
+
+// S3Target describes S3 target
+type S3Target struct {
+	logger log.Logger
+
+	handler   api.EntryHandler
+	positions positions.Positions
+	labels    model.LabelSet
+
+	watch map[string]*objectReader
+	path  string
+	quit  chan struct{}
+	done  chan struct{}
+
+	config       aws.S3Config
+	objectClient chunk.ObjectClient
+	prefix       string
+	syncPeriod   time.Duration
+}
+
+// NewFileTarget create a new FileTarget.
+func NewS3Target(logger log.Logger, handler api.EntryHandler, positions positions.Positions, jobName string, objectClient chunk.ObjectClient, s3Config *scrape.S3Targetconfig) (*S3Target, error) {
+	// // object store sync period should not be less than 1 minute
+	// if s3Config.SyncPeriod < 1*time.Minute {
+	// 	return nil, errors.New("SyncPeriod should not be less than 1 minute")
+	// }
+
+	// add prefix label
+	labels := s3Config.Labels.Merge(model.LabelSet{
+		model.LabelName("s3_prefix"): model.LabelValue(s3Config.Prefix),
+	})
+
+	t := &S3Target{
+		logger:       logger,
+		handler:      handler,
+		positions:    positions,
+		quit:         make(chan struct{}),
+		done:         make(chan struct{}),
+		labels:       labels,
+		objectClient: objectClient,
+		watch:        map[string]*objectReader{},
+		prefix:       s3Config.Prefix,
+		syncPeriod:   s3Config.SyncPeriod,
+	}
+
+	go t.run()
+	return t, nil
+}
+
+func (t *S3Target) run() {
+	defer func() {
+		for _, v := range t.watch {
+			level.Info(t.logger).Log("msg", "updating object last position", "object", v.object.Key)
+			v.markPositionAndSize()
+			level.Info(t.logger).Log("msg", "stopping object reader", "object", v.object.Key)
+			v.stop()
+		}
+		close(t.done)
+	}()
+
+	ticker := time.NewTicker(t.syncPeriod)
+	for {
+		select {
+		case <-ticker.C:
+			objects, _, err := t.objectClient.List(context.Background(), t.prefix)
+			if err != nil {
+				level.Error(t.logger).Log("msg", "failed to list objects", "error", err)
+				continue
+			}
+			t.handleObjects(objects)
+		case <-t.quit:
+			return
+		}
+	}
+}
+
+func (t *S3Target) handleObjects(objects []chunk.StorageObject) {
+	for _, object := range objects {
+		if v, ok := t.watch[object.Key]; ok {
+			if v.err == nil && v.active {
+				continue
+			}
+			level.Info(t.logger).Log("msg", "stopping object reader", "object", v.object.Key)
+			v.stop()
+			delete(t.watch, object.Key)
+			continue
+		}
+
+		// check if the object's position is already saved
+		modifiedAt, pos, size, err := t.getPosition(object.Key)
+		if err != nil {
+			level.Error(t.logger).Log("msg", "error in getting position for object", "error", err, "object", object.Key)
+			continue
+		}
+
+		// skip any object which is already read completely
+		if modifiedAt == object.ModifiedAt.UnixNano() && pos > 0 && pos == size {
```

Comment:
Do you need the `pos > 0` check? It seems unnecessary.

Reply:
There might be a chance that Promtail shutdowns without reading a single line. On restart pos and size will be `0`.So, just checking `pos == size` will ignore the object.